src/core/connection.c

```diff
@@ -4065,6 +4081,37 @@ QuicConnRecvFrames(
             break;
         }
 
+        case QUIC_FRAME_DATAGRAM:
+        case QUIC_FRAME_DATAGRAM_1: {
+            QUIC_DBG_ASSERT(Connection->State.PeerTransportParameterValid);
+            if (Connection->PeerTransportParams.MaxDatagramFrameSize == 0) {
+                QuicTraceEvent(
+                    ConnError,
+                    "[conn][%p] ERROR, %s.",
+                    Connection,
+                    "Received DATAGRAM frame when not negotiated");
+                QuicConnTransportError(Connection, QUIC_ERROR_PROTOCOL_VIOLATION);
+                return FALSE;
+            }
+            if (!QuicDatagramProcessFrame(
+                    &Connection->Datagram,
+                    Packet,
+                    FrameType,
+                    PayloadLength,
+                    Payload,
+                    &Offset)) {
+                QuicTraceEvent(
+                    ConnError,
+                    "[conn][%p] ERROR, %s.",
+                    Connection,
+                    "Decoding DATAGRAM frame");
+                QuicConnTransportError(Connection, QUIC_ERROR_FRAME_ENCODING_ERROR);
+                return FALSE;
+            }
+            AckPacketImmediately = TRUE;

```

Comment:
Do we need to ACK immediately? Or is the spec relaxed about this?

Reply:
Yes, it's ACK iliciting, so we need to acknowledge immediately.